oggm/core/models/flowline.py

```diff
@@ -623,6 +625,153 @@ def step(self, dt=SEC_IN_MONTH):
+            D_up_p = Gamma * H_up_p**(N+2.) * s_grad_up                  # Eq. 30
+            
+            D_up_min = np.minimum(D_up_m,D_up_p);                        # Eq. 31
+            D_up_max = np.maximum(D_up_m,D_up_p);                        # Eq. 32
+            D_up = np.zeros(fl.nx)
+            
+            # Eq. 33
+            D_up[np.logical_and(S[kp]<=S[k],H_up_m<=H_up_p)] = D_up_min[np.logical_and(S[kp]<=S[k],H_up_m<=H_up_p)]
+            D_up[np.logical_and(S[kp]<=S[k],H_up_m>H_up_p)] = D_up_max[np.logical_and(S[kp]<=S[k],H_up_m>H_up_p)]
+            D_up[np.logical_and(S[kp]>S[k],H_up_m<=H_up_p)] = D_up_max[np.logical_and(S[kp]>S[k],H_up_m<=H_up_p)]
+            D_up[np.logical_and(S[kp]>S[k],H_up_m>H_up_p)] = D_up_min[np.logical_and(S[kp]>S[k],H_up_m>H_up_p)]
+
+            # MUSCL scheme down. "down" denotes here the k-1/2 flux boundary
+            r_dn_m = (H[km]-H[kmm])/(H[k]-H[km])
+            H_dn_m = H[km] + 0.5 * self.phi(r_dn_m)*(H[k]-H[km])
+            r_dn_p = (H[k]-H[km])/(H[kp]-H[k])
+            H_dn_p = H[k] - 0.5 * self.phi(r_dn_p)*(H[kp]-H[k])
+            
+            # calculate the slope gradient
+            s_grad_dn = ((S[k]-S[km])**2. / dx**2.)**((N-1.)/2.)
+            D_dn_m = Gamma * H_dn_m**(N+2.) * s_grad_dn
+            D_dn_p = Gamma * H_dn_p**(N+2.) * s_grad_dn
+            
+            D_dn_min = np.minimum(D_dn_m,D_dn_p);
+            D_dn_max = np.maximum(D_dn_m,D_dn_p);
+            D_dn = np.zeros(fl.nx)
+            
+            D_dn[np.logical_and(S[k]<=S[km],H_dn_m<=H_dn_p)] = D_dn_min[np.logical_and(S[k]<=S[km],H_dn_m<=H_dn_p)]
+            D_dn[np.logical_and(S[k]<=S[km],H_dn_m>H_dn_p)] = D_dn_max[np.logical_and(S[k]<=S[km],H_dn_m>H_dn_p)]
+            D_dn[np.logical_and(S[k]>S[km],H_dn_m<=H_dn_p)] = D_dn_max[np.logical_and(S[k]>S[km],H_dn_m<=H_dn_p)]
+            D_dn[np.logical_and(S[k]>S[km],H_dn_m>H_dn_p)] = D_dn_min[np.logical_and(S[k]>S[km],H_dn_m>H_dn_p)]
+            
+            dt_stab = c_stab * dx**2. / max(max(abs(D_up)),max(abs(D_dn)))      # Eq. 37
+            dt_use = min(dt_stab,dt-stab_t)
+            stab_t = stab_t + dt_use
+            
+            # check if the extra time stepping is needed [to be removed one day]
+            #if dt_stab < dt:
+            #    print "MUSCL extra time stepping dt: %f dt_stab: %f" % (dt, dt_stab)
+            #else:
+            #    print "MUSCL Scheme fine with time stepping as is"
+            
+            #explicit time stepping scheme
+            div_q = (D_up * (S[kp] - S[k])/dx - D_dn * (S[k] - S[km])/dx)/dx    # Eq. 36
+            S = S[k] + (m_dot + div_q)*dt_use                                   # Eq. 35
+            
+            S = np.maximum(S,B)                                                 # Eq. 7
+        
+        # Done with the loop, prepare output
+        NewIceThickness = S-B
+        
+        fl.section = NewIceThickness * width
```

Comment:
Here you could set `fl.thick = NewIceThickness`

This would have the advantage that you don't have to assume a U-shaped glacier bed. If you have a look at the code for `ParabolicFlowline`, `VerticalWallFlowline`, and `TrapezoidalFlowline` (https://github.com/OGGM/oggm/blob/master/oggm/core/models/flowline.py#L89), you'll see that setting `fl.section` is equivalent to setting the thickness, which is the actual parameter defining the section. This is how the `FluxBasedModel` works, regardless of which shape the glacier bed has.

Note that in a way the `FluxBasedModel` physics are a bit wrong (in https://github.com/OGGM/oggm/blob/master/oggm/core/models/flowline.py#L476, the thickness is physically the same regardless of the bed shape). How "wrong" do you think that your maths above will be if the bed is e.g. parabolic? 

I hope that I've been clear, otherwise I'll try to elaborate more.